### ISSUE TITLE
Broken symbolic link causes crash and startup failure

### ISSUE THREAD

**Issue by DaveJarvis** [processing/processing#5977]
https://github.com/processing/processing/blob/d6b30602a703e1d0280392c54794486ee34bca8c/java/src/processing/mode/java/JavaEditor.java#L1561

If `code` is `null`, the next line will throw a NullPointerException.

**Replicate**

1. `mkdir $HOME/dev/processing/npe`
1. `cd $HOME/dev/processing/npe`
1. `echo "setup() {}" > npe.pde`
1. `touch Test.java`
1. `ln -s Test.java Test2.java`
1. `rm Test.java`
1. Start Processing
1. Open `npe.pde`

**Expected**

Broken symbolic link is ignored.

**Actual**

![npe](https://user-images.githubusercontent.com/2131950/71777459-189eb000-2f55-11ea-89ee-9bf7f1f8f334.png)

**Comment by Bobby54** [comment]
I translated it to Windows and it works, however it doesn't work on Linux Ubuntu.
What is your OS and Processing Version?
Do you have an error message in the console about not having permission to read the file or that the file has gone missing?

**Comment by DaveJarvis** [comment]
> I translated it to Windows and it works, however it doesn't work on Linux Ubuntu.

Try creating a write-only file (i.e., remove read permissions).

> What is your OS and Processing Version?

Processing v3.5.3 on Linux.

> Do you have an error message in the console about not having permission to read the file or that the file has gone missing?

Yes, there was another error message on the console, to the effect that the file could not be opened or found.

**Comment by Bobby54** [comment]
> Try creating a write-only file (i.e., remove read permissions).
I actually checked what happens if you change permissions which resulted in the same error.

**Comment by Bobby54** [comment]
I think you mustn't remove Test.java because npe is linked to it.
If I follow your steps to replicate it works fine if I don't remove Test.java. But if I do it returns a NullPointerException.

**Comment by DaveJarvis** [comment]
> I think you mustn't remove Test.java because npe is linked to it.

Should the software crash when it cannot open a missing/write-only file?

If no, then change 1562 to something like:

``` java
String lines[] = code == null ? new String[0] : code.split("\\r?\\n");
```

If yes, then close this ticket.

**Comment by Bobby54** [comment]
> Should the software crash when it cannot open a missing/write-only file?

No, it shouldn't, but think of a link on a website to a different website which doesn't exist anymore.
The software basically crashes but there is an output like "This website doesn't exist anymore".
So for this case there should be an output stating something like "File doesn't exist".
It would make more sense than opening an empty file.

**Comment by DaveJarvis** [comment]
> So for this case there should be an output stating something like "File doesn't exist".

Is that accurate enough? Something like, "File could not opened for reading" may be more helpful. (After all, a write-only file does exist, but cannot be read; similarly, a symbolic link exists, even though its contents may direct to a non-existent file.)

Either way, yes, having output that indicates to the user the list of files that could not be read would be useful. Or, even simpler, noting that "Some files could not be read" with a pop-up and direct the user to see console for details for the list of files that were not imported into the IDE.

**Comment by Bobby54** [comment]
I just figured out how to solve your issue.
Instead of `ln -s` type `ln`.
The difference is that `ln -s` creates a symbolic link which is deleted after deleting one of the files. `ln ` however creates a hard link which is also usable after deleting the file the link is pointing to.
If you try running everything again, you will see that the issue is solved.

**Comment by DaveJarvis** [comment]
> I just figured out how to solve your issue.

Thank you! Yes, that would have definitely _avoided_ the issue for me in this particular case. However, it doesn't address the underlying problem of a `NullPointerException` lurking in the code that other people can encounter when using write-only or missing files, which is readily resolved with a simple check.
